src/v/pandaproxy/schema_registry/protobuf.cc

```diff
@@ -499,14 +499,33 @@ struct compatibility_checker {
             }
         }
 
+        // check writer fields
         for (int i = 0; i < writer->field_count(); ++i) {
+            // TODO(oren): Need special handling here?
             if (reader->IsReservedNumber(i) || writer->IsReservedNumber(i)) {
                 continue;
             }
-            int number = writer->field(i)->number();
+            auto w = writer->field(i);
+            int number = w->number();
             auto r = reader->FindFieldByNumber(number);
-            // A reader may ignore a writer field
-            if (r && !check_compatible(r, writer->field(i))) {
+            // A reader may ignore a writer field iff it is not `required`
+            if (!r && w->is_required()) {
+                return false;
+            } else if (r && !check_compatible(r, w)) {
+                return false;
+            }
```

Comment:
nitpick: maybe

```suggestion
            return (!r && w->is_required()) || (r && !check_compatible(r, w));
```

Since we're short circuiting the whole function. Though maybe that's even harder to read!

Reply:
Yeah I think the existing one is easier to reason about. (I think this one is also missing a negation from the front to be equivalent with the existing code.)